File: 0748. Shortest Completing Word/0748.java
// time complexity: O(W·M·L) - worst case. where:
//      - W is number of the words
//      - M is length of the each word
//      - L is length of the "characters" variable
// space complexity: O(L)
class Solution {
    public String shortestCompletingWord(String licensePlate, String[] words) {
        // get only letters from licensePlate, and convert to lowercase
        StringBuilder characters = new StringBuilder();
        for (char c : licensePlate.toLowerCase().toCharArray()) {
            if (c >= 'a' && c <= 'z') {
                characters.append(c);
            }
        }

        String ans = null;
        for (String w : words) {
            // skip if there is already a shorter
            if (ans != null && w.length() >= ans.length()) continue;

            // create copy of characters content
            StringBuilder need = new StringBuilder(characters);

            // try to delete one from need for each letter in the word
            for (char c : w.toCharArray()) {
                int index = need.indexOf(String.valueOf(c)); // c is char, so use "valueOf()"
                if (index != -1) { // indexOf() returns -1, if doesn't exist
                    need.deleteCharAt(index);
                }
                if (need.length() == 0) break;  // break, if all of the letters are matched
            }

            // valid answer, if need is empty
            if (need.length() == 0) {
                ans = w;
            }
        }

        return ans;
    }
}

File: 0748. Shortest Completing Word/0748.py
class Solution(object):
    def shortestCompletingWord(self, licensePlate, words):
        """
        :type licensePlate: str
        :type words: List[str]
        :rtype: str
        """
        # 1) Count letters in licensePlate (convert to lowercase, ignore digits/spaces)
        lp_count = {}
        for ch in licensePlate.lower():
            if ch.isalpha(): # check if it's letter or not
                lp_count[ch] = lp_count.get(ch, 0) + 1

        answer = None

        # 2) Check each word in the list
        for w in words:
            # skip if we already have a shorter or equal answer
            if answer is not None and len(w) >= len(answer):
                continue

            # 3) Count letters in the current word
            w_count = {}
            for ch in w:
                w_count[ch] = w_count.get(ch, 0) + 1

            # 4) Verify the word covers all required letters
            valid = True
            for letter, req in lp_count.items():
                if w_count.get(letter, 0) < req:
                    valid = False
                    break

            # 5) If valid, update the answer
            if valid:
                answer = w

        # 6) Return the found answer (guaranteed to exist)
        return answer